src/smpp_socket.erl

```diff
@@ -955,3 +955,41 @@ pp(submit_sm_resp, _) -> record_info(fields, submit_sm_resp);
 pp(unbind, _) -> record_info(fields, unbind);
 pp(unbind_resp, _) -> record_info(fields, unbind_resp);
 pp(_, _) -> no.
+
+%%%-------------------------------------------------------------------
+%%% Overload protection
+%%%-------------------------------------------------------------------
+-spec init_rate_limit(state()) -> state().
+init_rate_limit(#{req_per_sec := undefined} = State) ->
+    State;
+init_rate_limit(#{req_per_sec := Rps} = State) when is_integer(Rps) ->
+    State#{current_rps := {erlang:monotonic_time(second), 0}}.
+
+-spec check_limits(state()) ->
+          {ok, state()} |
+          {error, {statename(), state(), [gen_statem:action()]}}.
+check_limits(#{req_per_sec := undefined} = State) ->
+    check_in_flight(State);
+check_limits(#{req_per_sec := MaxRps, current_rps := {CurSecond, CurRate}} = State) ->
+    case erlang:monotonic_time(second) of
+        CurSecond when CurRate == MaxRps ->
+            Now = erlang:monotonic_time(millisecond),
+            Timeout = (CurSecond + 1) * 1000 - Now,
```

Comment:
We still need to make sure, that time has't wrapped to the next second:
```suggestion
            Timeout = max(0, (CurSecond + 1) * 1000 - Now),
```

Reply:
fixed